torch/lib/c10d/test/TCPStoreTest.cpp

```diff
@@ -136,10 +151,221 @@ void testHelper(const std::string& prefix = "") {
   }
 }
 
+void testWatchKeyCallback(const std::string& prefix = "") {
+  // Callback function increments counter of the total number of callbacks that
+  // were run
+  std::promise<int> numCallbacksExecutedPromise;
+  std::atomic<int> numCallbacksExecuted{0};
+  const int numThreads = 16;
+  const int keyChangeOperation = 3;
+  c10d::StoreCallbackFunction callback =
+      [&numCallbacksExecuted,
+       &numCallbacksExecutedPromise,
+       &numThreads,
+       &keyChangeOperation](
+          c10::optional<std::string> /* unused */,
+          c10::optional<std::string> /* unused */) {
+        numCallbacksExecuted++;
+        if (numCallbacksExecuted == numThreads * keyChangeOperation) {
+          numCallbacksExecutedPromise.set_value(numCallbacksExecuted);
+        }
+      };
+
+  const int numWorkers = numThreads + 1;
+  auto serverTCPStore = _createServer(numWorkers);
+  auto serverStore =
+      c10::make_intrusive<c10d::PrefixStore>(prefix, serverTCPStore);
+
+  // Start watching key
+  std::string internalKey = "internalKey";
+  for (auto i = 0; i < numThreads; i++) {
+    serverStore->watchKey(internalKey + std::to_string(i), callback);
+    serverStore->watchKey(
+        internalKey + "counter" + std::to_string(i), callback);
```

Comment:
IIUC, this means that all callbacks are installed by the server. This test mainly covers the case where clients modifies keys and server callbacks are invoked correctly. Am I correct?

Reply:
Yes all the callbacks are added to the process hosting the server. That is a good point, I will modify this test to also register callbacks on all clients and server.